feat!: Require generation in the multi-keyring (#157)
awslabs/aws-encryption-sdk-specification: changes/2020-07-14_multi-keyring-require-generation/change.md

[//]: # "Copyright Amazon.com Inc. or its affiliates. All Rights Reserved."
[//]: # "SPDX-License-Identifier: CC-BY-SA-4.0"

# Require Generation in the Multi-Keyring

## Affected Features

This serves as a reference of all features that this change affects.

| Feature                                                                                           |
| ------------------------------------------------------------------------------------------------- |

> | [Multi-Keyring](../../framework/multi-keyring.md)                                                 |

Reviewer: Quick extra note. I noticed in another PR you had the specific SHA the feature it changed as opposed to just the link. You should probably use the SHA here and everywhere else multi-keyring.md is linked

Author: Agreed, although whether the following section needs it as well is still up in the air: https://github.com/awslabs/aws-encryption-sdk-specification/pull/154#discussion_r455213501

> | [Multi keyring generator](https://github.com/awslabs/aws-encryption-sdk-specification/issues/114) |

Reviewer: np: we should stick with either `multi-keyring` or `multi keyring`

Author: Agreed, the second was copied from the issue title. Will fix.

| ---------- | ------------------------------------------------------------------------------------- |
| C          | [aws-encryption-sdk-c](https://github.com/aws/aws-encryption-sdk-c)                   |
| Javascript | [aws-encryption-sdk-javascript](https://github.com/aws/aws-encryption-sdk-javascript) |

This change additionally affects the unreleased AWS KMS keyring implementations for Java and Python.

## Definitions

### Conventions used in this document

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL"
in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).

## Summary

The multi-keyring has a clearly configured intent of which keyring SHOULD be used to generate the data key,
but this is not actually enforced during the encryption operation.
This change strengthens the requirements on multi-keyrings such that if a generator is configured,
it MUST generate the data key.

## Out of Scope

Detecting at keyring configuration time that encryption will always fail

> because keyrings are not capable of meeting these requirements

Reviewer: Might be worth saying what "these" requirements are. Is it "because the generator keyring cannot call AWS KMS GenerateDataKey or already has a plaintext data key" or is there something more?

Author: Left over from an earlier version that had at least one more requirement (that turned out to be already specified). But the general concern is still more broad so I'll indicate that.

rather than being certain that it will always do the same thing no matter what.

## Drawbacks

This is a breaking change,
since existing multi-keyring configurations
can be used in multiple contexts,
and data key generation is not tightly controlled.
We judge this to be the right trade-off
between security and usability.

## Security Implications

Removing this flexibility may help identify potential security issues,
as customers may discover when upgrading their ESDK version
that data keys are not being generated as intended.
Correcting this will allow customers
to improve their security postures
through improvements such as scoping down permissions.

## Operational Implications

Customers that are currently using a single multi-keyring instance
in multiple contexts with inconsistent data key generation patterns
will encounter errors when upgrading.
This may lead to an increase in support requests,
but should also reduce troubleshooting requests
related to unexpected behaviour around data key generation,
or unexpected permissions errors due to the same.

## Guide-level Explanation

The description of OnEncrypt for the multi-keyring
will be changed to read as follows:

> If this keyring has a generator keyring,
> this keyring MUST first generate a plaintext data key using the generator keyring.
> If the input encryption materials already include a plaintext data key,
> OnEncrypt MUST fail.
> Otherwise,
> this keyring MUST first call that generator keyring's OnEncrypt
> using the input encryption materials as input.
> If the generator keyring fails OnEncrypt,
> this OnEncrypt MUST also fail.

Reviewer 2: ```suggestion
If this keyring has a generator keyring,
and the input encryption materials includes a plaintext data key,
OnEncrypt MUST fail.
This keyring MUST first call that generator keyring's OnEncrypt
using the input encryption materials as input.
If the generator keyring fails OnEncrypt,
this OnEncrypt MUST also fail.
```

Author: Ended up using bullet points to clarify the scoping here instead

in this document are to be interpreted as described in [RFC 2119](https://tools.ietf.org/html/rfc2119).

## Summary

The multi-keyring has a clearly configured intent of which keyring SHOULD be used to generate the data key,
but this is not actually enforced during the encryption operation.
This change strengthens the requirements on multi-keyrings such that if a generator is configured,
it MUST generate the data key.

## Out of Scope

If a multi-keyring is configured with another multi-keyring as its generator,
and this second multi-keyring has no generator,
the outer multi-keyring cannot ever successfully complete an OnEncrypt operation.
This is one of several examples of keyring configurations
that cannot satisfy the requirements of an operation.
Detecting at keyring configuration time that encryption/decryption will always fail
because keyrings are not capable of meeting their requirements
is [tracked separately](https://github.com/awslabs/aws-encryption-sdk-specification/issues/144).

## Motivation

If we do not make a guarantee that the generator will generate,
then its usefulness is limited.

> Customers may have a compliance requirement that all data keys MUST be generated by a specific HSM,

Reviewer 3: wording: Avoid "may" when it is not "MAY". In this case I would use "might", "could", or similar.

or may want to lock down true least-privilege permissions for their AWS KMS key policies.
Under the current model,
customers have to be very careful to make sure that they know
exactly where and how a given multi-keyring configuration is being used
rather than being certain that it will always do the same thing no matter what.

## Drawbacks

This is a breaking change,
since existing multi-keyring configurations
can be used in multiple contexts,
and data key generation is not tightly controlled.
We judge this to be the right trade-off
between security and usability.

## Security Implications

Removing this flexibility may help identify potential security issues,
as customers may discover when upgrading their ESDK version
that data keys are not being generated as intended.
Correcting this will allow customers
to improve their security postures

> through improvements such as scoping down permissions.

Reviewer 3: I would split this into two separate implications:

1. Do the correct thing.
1. Scope down permissions.

## Operational Implications

Customers that are currently using a single multi-keyring instance
in multiple contexts with inconsistent data key generation patterns
will encounter errors when upgrading.

> This may lead to an increase in support requests,

Reviewer 3: ```suggestion
This might lead to an increase in support requests,
```

> but should also reduce troubleshooting requests

Reviewer 3: ```suggestion
but SHOULD also reduce troubleshooting requests
```

related to unexpected behaviour around data key generation,
or unexpected permissions errors due to the same.

## Guide-level Explanation

The description of OnEncrypt for the multi-keyring
will be changed to read as follows:

If this keyring has a generator keyring,
this keyring MUST first generate a plaintext data key using the generator keyring:

- If the input encryption materials already include a plaintext data key,
  OnEncrypt MUST fail.

> - Otherwise,
>   this keyring MUST first call that generator keyring's OnEncrypt
>   using the input encryption materials as input.
>   If the generator keyring fails OnEncrypt,
>   this OnEncrypt MUST also fail.
>   If the generator keyring returns encryption materials missing a plaintext data key,
>   OnEncrypt MUST fail.

Reviewer 3: I would split this into individual requirements.

```suggestion
- This keyring MUST first call the generator keyring's OnEncrypt
  using the input encryption materials as input.
- If the generator keyring fails OnEncrypt,
  this OnEncrypt MUST also fail.
- If the generator keyring returns encryption materials missing a plaintext data key,
  OnEncrypt MUST fail.
```

If we do not make a guarantee that the generator will generate,
then its usefulness is limited.
Customers might have a compliance requirement that all data keys MUST be generated by a specific HSM,
or might want to lock down true least-privilege permissions for their AWS KMS key policies.
Under the current model,

> users have to be very careful to make sure that they know

Reviewer: ```suggestion
customers have to be very careful to make sure that they know
```

## Security Implications

Removing this flexibility might help identify potential security issues,
as users might discover when upgrading their ESDK version

> that data keys are not being generated as intended.

Reviewer: ```suggestion
Removing this flexibility might help identify potential security issues.
For example, when upgrading their version of the ESDK, customers might discover
that data keys are not being generated as intended.
```

Author: Thanks, applied this manually
